### ISSUE TITLE
test-errors related to FileUpload in Plone 5.2

### ISSUE THREAD

**Issue by pbauer** [plone/Products.CMFPlone#2193]
Two test related to FileUpload fail.

1. `./bin/test -s plone.formwidget.namedfile -t widget.rst`. Traceback: http://jenkins.plone.org/job/plone-5.2-python-2.7/61/testReport/junit/plone.formwidget/namedfile/widget_rst/ 
2. `./bin/test -s plone.app.blob -t replacement-types.txt`. Traceback: http://jenkins.plone.org/job/plone-5.2-python-2.7-at/56/testReport/junit/src.plone.app.blob/tests/replacement_types_txt/

I think both are related to the patches in `plone.app.blob.monkey` being moved into Zope with https://github.com/zopefoundation/Zope/commit/4d5910f3130dbddd4 by @hannosch but I could not track down the real reason. 

In the `replacement-types.txt` test the whole `TemporaryFileWrapper` is not used at all and I cannot figure our why. The tested feature (changing the image of a ATBlob Image object) works in "real life".

**Comment by mauritsvanrees** [comment]
The formwidget errors are because we are mocking FileUpload and the mock needs to incorporate some changes to the original class. Done in:
- https://github.com/plone/plone.formwidget.namedfile/commit/b6c5d4e434920b3be7cdd5cc4447802b5302c33e
- https://github.com/plone/plone.formwidget.namedfile/commit/5510211f90ae46d866268b03144bb2e7641633a3
- https://github.com/plone/plone.formwidget.namedfile/commit/c6cb007cbd59d8dae31aa0b62437cd865aed363a

**Comment by mauritsvanrees** [comment]
It looks like the monkey patch in `plone.app.blob` could be removed, if we split off a branch for Plone 5.0 and 5.1.  But when I remove the patch, the failures stay the same.

**Comment by pbauer** [comment]
TL;DR Uploading gif images is broken in Plone 5.2.

So I thought the reason why `TemporaryFileWrapper` was not used had something to do with it being a doctest using PloneTestCase and created a new 'modern' that is easier to debug. But that failed in the same way:

```
Failure in test test_replace_blobimage (plone.app.blob.tests.test_integration.TestBlobBrokenTest)
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/Users/pbauer/workspace/coredev/src/plone.app.blob/src/plone/app/blob/tests/test_integration.py", line 337, in test_replace_blobimage
    self.browser.getControl('Save').click()
  File "/Users/pbauer/.cache/buildout/eggs/zope.testbrowser-5.2.3-py2.7.egg/zope/testbrowser/browser.py", line 771, in click
    self.browser._clickSubmit(self._form, self._control)
  File "/Users/pbauer/.cache/buildout/eggs/zope.testbrowser-5.2.3-py2.7.egg/zope/testbrowser/browser.py", line 275, in _clickSubmit
    self._processRequest(url, make_request)
  File "/Users/pbauer/.cache/buildout/eggs/zope.testbrowser-5.2.3-py2.7.egg/zope/testbrowser/browser.py", line 280, in _processRequest
    resp = make_request(reqargs)
  File "/Users/pbauer/.cache/buildout/eggs/zope.testbrowser-5.2.3-py2.7.egg/zope/testbrowser/browser.py", line 270, in make_request
    form, control.name, index, coord=coord, **args)
  File "/Users/pbauer/.cache/buildout/eggs/zope.testbrowser-5.2.3-py2.7.egg/zope/testbrowser/browser.py", line 310, in _submit
    params=fields, **args)
  File "/Users/pbauer/.cache/buildout/eggs/WebTest-2.0.28-py2.7.egg/webtest/response.py", line 297, in goto
    return method(href, **args)
  File "/Users/pbauer/.cache/buildout/eggs/WebTest-2.0.28-py2.7.egg/webtest/app.py", line 379, in post
    content_type=content_type)
  File "/Users/pbauer/.cache/buildout/eggs/WebTest-2.0.28-py2.7.egg/webtest/app.py", line 755, in _gen_request
    expect_errors=expect_errors)
  File "/Users/pbauer/.cache/buildout/eggs/zope.testbrowser-5.2.3-py2.7.egg/zope/testbrowser/browser.py", line 94, in do_request
    expect_errors)
  File "/Users/pbauer/.cache/buildout/eggs/WebTest-2.0.28-py2.7.egg/webtest/app.py", line 625, in do_request
    res = req.get_response(app, catch_exc_info=True)
  File "/Users/pbauer/.cache/buildout/eggs/WebOb-1.7.3-py2.7.egg/webob/request.py", line 1312, in send
    application, catch_exc_info=True)
  File "/Users/pbauer/.cache/buildout/eggs/WebOb-1.7.3-py2.7.egg/webob/request.py", line 1280, in call_application
    app_iter = application(self.environ, start_response)
  File "/Users/pbauer/.cache/buildout/eggs/WebTest-2.0.28-py2.7.egg/webtest/lint.py", line 199, in lint_app
    iterator = application(environ, start_response_wrapper)
  File "/Users/pbauer/workspace/coredev/src/plone.testing/src/plone/testing/_z2_testbrowser.py", line 36, in wrapped_func
    return func(*args, **kw)
  File "/Users/pbauer/workspace/coredev/src/plone.testing/src/plone/testing/_z2_testbrowser.py", line 63, in __call__
    wsgi_result = publish(environ, start_response)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/ZPublisher/WSGIPublisher.py", line 256, in publish_module
    response = _publish(request, new_mod_info)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/ZPublisher/WSGIPublisher.py", line 209, in publish
    bind=1)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/ZPublisher/mapply.py", line 85, in mapply
    return debug(object, args, context)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/ZPublisher/WSGIPublisher.py", line 56, in call_object
    return obj(*args)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFFormController/Products/CMFFormController/FSControllerPageTemplate.py", line 91, in __call__
    return self._call(FSControllerPageTemplate.inheritedAttribute('__call__'), *args, **kwargs)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFFormController/Products/CMFFormController/BaseControllerPageTemplate.py", line 29, in _call
    return self.getNext(controller_state, REQUEST)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFFormController/Products/CMFFormController/ControllerBase.py", line 232, in getNext
    return next_action.getAction()(controller_state)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFFormController/Products/CMFFormController/Actions/TraverseTo.py", line 38, in __call__
    REQUEST, bind=1)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/ZPublisher/mapply.py", line 85, in mapply
    return debug(object, args, context)
  File "/Users/pbauer/.cache/buildout/eggs/ZServer-4.0b1-py2.7.egg/ZServer/ZPublisher/Publish.py", line 44, in call_object
    return object(*args)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFFormController/Products/CMFFormController/FSControllerPythonScript.py", line 110, in __call__
    return self.getNext(result, request)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFFormController/Products/CMFFormController/ControllerBase.py", line 232, in getNext
    return next_action.getAction()(controller_state)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFFormController/Products/CMFFormController/Actions/TraverseTo.py", line 38, in __call__
    REQUEST, bind=1)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/ZPublisher/mapply.py", line 85, in mapply
    return debug(object, args, context)
  File "/Users/pbauer/.cache/buildout/eggs/ZServer-4.0b1-py2.7.egg/ZServer/ZPublisher/Publish.py", line 44, in call_object
    return object(*args)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFFormController/Products/CMFFormController/FSControllerPythonScript.py", line 108, in __call__
    result = FSControllerPythonScript.inheritedAttribute('__call__')(self, *args, **kwargs)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFFormController/Products/CMFFormController/Script.py", line 145, in __call__
    return BaseFSPythonScript.__call__(self, *args, **kw)
  File "/Users/pbauer/workspace/coredev/src/Products.CMFCore/Products/CMFCore/FSPythonScript.py", line 126, in __call__
    return Script.__call__(self, *args, **kw)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/Shared/DC/Scripts/Bindings.py", line 330, in __call__
    return self._bindAndExec(args, kw, None)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/Shared/DC/Scripts/Bindings.py", line 367, in _bindAndExec
    return self._exec(bound_data, args, kw)
  File "/Users/pbauer/workspace/coredev/src/Products.PythonScripts/src/Products/PythonScripts/PythonScript.py", line 354, in _exec
    result = function(*args, **kw)
  File "Script (Python)", line 1, in content_edit
   - Line 1
  File "/Users/pbauer/workspace/coredev/src/Products.CMFCore/Products/CMFCore/FSPythonScript.py", line 126, in __call__
    return Script.__call__(self, *args, **kw)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/Shared/DC/Scripts/Bindings.py", line 330, in __call__
    return self._bindAndExec(args, kw, None)
  File "/Users/pbauer/workspace/coredev/src/Zope/src/Shared/DC/Scripts/Bindings.py", line 367, in _bindAndExec
    return self._exec(bound_data, args, kw)
  File "/Users/pbauer/workspace/coredev/src/Products.PythonScripts/src/Products/PythonScripts/PythonScript.py", line 354, in _exec
    result = function(*args, **kw)
  File "Script (Python)", line 12, in content_edit_impl
   - Line 12
  File "/Users/pbauer/workspace/coredev/src/Products.Archetypes/Products/Archetypes/BaseObject.py", line 675, in processForm
    REQUEST=REQUEST, values=values)
  File "/Users/pbauer/workspace/coredev/src/Products.Archetypes/Products/Archetypes/BaseObject.py", line 665, in _processForm
    mapply(mutator, result[0], **result[1])
   - __traceback_info__: (<plone.app.blob.content.ATBlob object at 0x10dac8500>, <Field image(blob:rw)>, <function mutator at 0x10d7da6e0>)
  File "/Users/pbauer/workspace/coredev/src/Products.Archetypes/Products/Archetypes/utils.py", line 131, in mapply
    return method(*call_args, **kw)
  File "/Users/pbauer/.cache/buildout/eggs/archetypes.schemaextender-2.1.7-py2.7.egg/archetypes/schemaextender/field.py", line 39, in mutator
    self.set(instance, value, **kw)
  File "/Users/pbauer/workspace/coredev/src/plone.app.blob/src/plone/app/blob/subtypes/image.py", line 22, in set
    super(ExtensionBlobField, self).set(instance, value, **kwargs)
  File "/Users/pbauer/workspace/coredev/src/plone.app.blob/src/plone/app/blob/field.py", line 249, in set
    blobbable.feed(blob.getBlob())
  File "/Users/pbauer/workspace/coredev/src/plone.app.blob/src/plone/app/blob/adapters/fileupload.py", line 37, in feed
    assert isfile(filename), 'invalid file for blob: {0}'.format(filename)  # noqa
AssertionError: invalid file for blob: image_file
```

After some more digging I suddenly had a passing test. WHAT??? So it turns out the issue is that it only fails with gif images, not jpg or png. I then tried to do exactly that in a real instance and it turns out that adding images with gif's is really broken in Plone 5.2. The file-extension does not matter, it is the actual image-type. I will fix the existing test to use a type that works and will add a failing test with a gif.

**Comment by pbauer** [comment]
The monkey patches could go away now but that is different task.

**Comment by mauritsvanrees** [comment]
Ouch. I guess it is better upload png or jpeg nowadays, but gif should still be possible.
